File: Stack/155.Min-Stack/155.Min-Stack.ts
class MinStack {
    private stack: number[]
    constructor() {
        this.stack = [];
    }

    push(val: number): void {
        if (this.stack.length === 0) {
            this.stack.push(val);
            this.stack.push(val);
        } else {
            const currentMin = this.getMin();
            this.stack.push(val);
            if (val < currentMin) {
                this.stack.push(val);
            } else {
                this.stack.push(currentMin);
            }
        }
    }

    pop(): void {
        this.stack.pop();
        this.stack.pop();
    }

    top(): number {
        return this.stack[this.stack.length - 2];
    }

    getMin(): number {
        return this.stack[this.stack.length - 1];
    }
}

File: Stack/155.Min-Stack/155.Min-Stack.py
class MinStack:

    def __init__(self):
        self.stack = []

    def push(self, val: int) -> None:
        currentMin = val
        if len(self.stack) != 0 and self.stack[-1] < val:
            currentMin = self.stack[-1]
        self.stack.append(val)
        self.stack.append(currentMin)

    def pop(self) -> None:
        self.stack.pop()
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-2]

    def getMin(self) -> int:
        return self.stack[-1]



# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
